src/uct/ib/mlx5/dv/ib_mlx5dv_md.c

```diff
@@ -508,6 +534,38 @@ uct_ib_mlx5_devx_dereg_mt(uct_ib_mlx5_md_t *md,
     return status;
 }
 
+static void uct_ib_mlx5_devx_reg_symmetric(uct_ib_mlx5_md_t *md,
+                                           uct_ib_mlx5_devx_mem_t *memh,
+                                           void *address)
+{
+    static const uint32_t size = 8;
+    struct mlx5dv_devx_obj *smkey_mr;
+    uint32_t symmetric_rkey;
+    ucs_status_t status;
+
+    /* Best effort, only allocate in the range below the atomic keys. */
+    while (md->symmetric_index < uct_ib_mlx5_symmetric_rkey_offset(md)) {
+        status = uct_ib_mlx5_devx_reg_ksm_data_contig(
+                md, &memh->mrs[UCT_IB_MR_DEFAULT], address, (uint64_t)address,
+                (memh->super.flags & UCT_IB_MEM_ACCESS_REMOTE_ATOMIC),
+                md->super.mkey_by_name_reserve.base + md->symmetric_index,
+                "symmetric key", &smkey_mr, &symmetric_rkey);
+        if (status == UCS_OK) {
+            memh->smkey_mr   = smkey_mr;
+            memh->super.rkey = symmetric_rkey;
+
+            md->symmetric_index++;
+            return;
+        }
+
+        /* Use blocks of 8 mkeys, first mkey creation gives block ownership.
+         * Try from the start of the next block if any failure.
+         */
```

Comment:
is this useful because of multiple segments?

Reply:
yes only for that, with a worst case of 32 segments, it would only be 32/8 create_mkey that contend between ranks on a node.